zebra-consensus/src/checkpoint.rs

```diff
@@ -155,4 +152,231 @@ where
     }
 }
 
-// TODO(teor): tests
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use color_eyre::Report;
+    use eyre::{bail, ensure, eyre};
+    use tower::{util::ServiceExt, Service};
+
+    use zebra_chain::serialization::ZcashDeserialize;
+
+    #[tokio::test]
+    #[spandoc::spandoc]
```

Comment:
Just a heads up on `spandoc`, this is a proc macro that converts doc comments _in the function_ into spans that instrument the line they're annotating. So with this function as written this proc macro has no effect, you'd need to turn the comments on line 172, 182, etc to be doc comments `///` to get it to trigger, which would cause errors thrown by those lines to include the comment as a span in the `SpanTrace` captured by that `eyre::Report`.

Another cool bonus fact of `spandoc` is that if the expression you're instrumenting has an `await` in it it will instead use `tracing-futures` and `.instrument(span)` to appropriately instrument the async call, and if the expression you're instrumenting has two awaits in it `spandoc` will throw a compiler error. Also any and all of this can be changed, if you have ideas for things you want to be able to write with spandoc lmk and I can add them.

Reply:
Fixed in #538 in 82de5b442796f8947d59af7f13d14fa02f77e298 for verify/block (but not checkpoint)

Reply:
Fixed in #538 in af50ec8434d706f4c155152a70c61380ec33998b for checkpoint

Reply:
> Also any and all of this can be changed, if you have ideas for things you want to be able to write with spandoc lmk and I can add them.

If I write a multi-line doc comment, I get a lint about an unused doc comment. And I assume that the "unused" doc comment doesn't end up in the span.

I would expect that multi-line doc comments would be turned into a span string. (Ideally multi-line, but the most important thing is that the comment isn't lost.)